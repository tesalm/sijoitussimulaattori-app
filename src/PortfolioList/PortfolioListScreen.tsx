import React from 'react';
import { ActivityIndicator, FlatList, Text, View } from 'react-native';
import { ListItem, SearchBar } from 'react-native-elements';
import { NavigationScreenProps } from 'react-navigation';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { t } from '../assets/i18n';
import { RootState } from '../redux/reducers';
import { getPortfolios, SaveAsCurrentPortfolioId } from './actions';
import { SinglePortfolio } from './reducers';

export interface PortfolioListProps {
  portfolios: Array<SinglePortfolio>;
  loading: boolean;
  error?: Error;
  getAllPortfolios: typeof getPortfolios;
  saveAsCurrentId: typeof SaveAsCurrentPortfolioId;
  portfolioId?: string;
}

type PortfolioPropsWithNavigation = PortfolioListProps & NavigationScreenProps;

export class PortfolioListScreen extends React.Component<
  PortfolioPropsWithNavigation
> {
  constructor(props: PortfolioPropsWithNavigation) {
    super(props);
  }

  componentDidMount() {
    //Dispatch the actions
    this.props.getAllPortfolios();
  }

  renderHeader = (): JSX.Element => {
    return (
      <SearchBar
        lightTheme
        round
        placeholder={t('ListStockPage.SearcBarPlaceholder')}
        //TODO: search bar functionality
        autoCorrect={false}
      />
    );
  };

  render() {
    const { portfolios, loading, error } = this.props;
    if (error) {
      //TODO: Format the error message to user
      return <Text>Error! {error.message} </Text>;
    }
    if (loading) {
      return (
        <View>
          <ActivityIndicator size="large" />
        </View>
      );
    }

    return (
      <FlatList
        data={portfolios}
        keyExtractor={(item) => item.name}
        renderItem={({ item, index }) => (
          <ListItem
            //TODO: navigate to to right stock page.
            onPress={() => {
              this.props.saveAsCurrentId(item.uid);
              this.props.navigation.navigate('SinglePortfolio');
            }}
            title={item.name}
          />
        )}
        ListHeaderComponent={this.renderHeader}
      />
    );
  }
}
const mapStateToProps = (state: RootState) => ({
  portfolios: state.portfolioListing.portfolios,
  loading: state.portfolioListing.loading,
  error: state.portfolioListing.error,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      getAllPortfolios: getPortfolios,
      saveAsCurrentId: SaveAsCurrentPortfolioId,
    },
    dispatch
  );
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PortfolioListScreen);
