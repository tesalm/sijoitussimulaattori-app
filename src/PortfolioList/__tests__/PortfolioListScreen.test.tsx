import * as React from 'react';
import renderer from 'react-test-renderer';

import { PortfolioListProps, PortfolioListScreen } from '../PortfolioListScreen';

describe('portfolio-list tests', () => {
  //Mock for navigation.
  const navigationMock: any = {};

  const defaultPortfolioListingProps: PortfolioListProps = {
    portfolioListing: [],
    loading: false,

    error: undefined,
    getAllPortfolios: jest.fn(),
    saveAsCurrentId: jest.fn(),
  };

  const loadingPortfolioListingProps: PortfolioListProps = {
    portfolioListing: [],
    loading: true,

    error: undefined,
    getAllPortfolios: jest.fn(),
    saveAsCurrentId: jest.fn(),
  };

  const errorPortfolioListingProps: PortfolioListProps = {
    portfolioListing: [],
    loading: false,

    error: { name: 'Network Error', message: 'Network connection failed' },
    getAllPortfolios: jest.fn(),
    saveAsCurrentId: jest.fn(),
  };

  const portfolioPortfolioListingProps: PortfolioListProps = {
    portfolioListing: [
      {
        uid: 'Portfolio1',
        name: 'Portfolio1',
        balance: 500,
        ownerId: 'Jooni',
        totalRevenue: 0.25,
        totalMarketValue: 0.53,
        lastDayRevenue: 0.23,
        portfolioInfo: {
          loading: false,
          error: undefined,
          refreshing: false,
          portfolio: {
            uid: 'Portfolio1',
            name: 'Portfolio1',
            balance: 500,
            totalRevenue: -0.02,
            totalMarketValue: 5000,
            lastDayRevenue: 0,
            ownerId: 'Joooni',
            stocks: [
              {
                uid: 'AMZN',
                amount: 245,
                avgPrice: 153.3,
                totalRevenue: 0.54,
                totalMarketValue: 2084,
                lastDayRevenue: 0.52,
              },
              {
                uid: 'AAPL',
                amount: 340,
                avgPrice: 132,
                totalRevenue: 0.54,
                totalMarketValue: 2084,
                lastDayRevenue: 0.52,
              },
            ],
          },
        },
      },
    ],

    loading: false,
    error: undefined,
    getAllPortfolios: jest.fn(),
    saveAsCurrentId: jest.fn(),
  };

  it('renders correctly', async () => {
    const component = renderer
      .create(
        <PortfolioListScreen
          {...defaultPortfolioListingProps}
          {...navigationMock}
        />
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with loading', async () => {
    const component = renderer
      .create(
        <PortfolioListScreen
          {...loadingPortfolioListingProps}
          {...navigationMock}
        />
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with error', async () => {
    const component = renderer
      .create(
        <PortfolioListScreen
          {...errorPortfolioListingProps}
          {...navigationMock}
        />
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with portfolio', async () => {
    const component = renderer
      .create(
        <PortfolioListScreen
          {...portfolioPortfolioListingProps}
          {...navigationMock}
        />
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
