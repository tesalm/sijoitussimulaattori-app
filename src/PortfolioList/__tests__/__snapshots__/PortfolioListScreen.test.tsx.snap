// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`portfolio-list tests renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#004D40",
        "fontSize": 40,
        "textAlign": "center",
      }
    }
  >
    You don't have any portfolios
  </Text>
</View>
`;

exports[`portfolio-list tests renders correctly with error 1`] = `
<Text>
  Network connection failed
</Text>
`;

exports[`portfolio-list tests renders correctly with loading 1`] = `
<View>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`portfolio-list tests renders correctly with portfolio 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  data={
    Array [
      Object {
        "balance": 500,
        "lastDayRevenue": 0.23,
        "name": "Portfolio1",
        "ownerId": "Jooni",
        "portfolioInfo": Object {
          "error": undefined,
          "loading": false,
          "portfolio": Object {
            "balance": 500,
            "lastDayRevenue": 0,
            "name": "Portfolio1",
            "ownerId": "Joooni",
            "stocks": Array [
              Object {
                "amount": 245,
                "avgPrice": 153.3,
                "lastDayRevenue": 0.52,
                "totalMarketValue": 2084,
                "totalRevenue": 0.54,
                "uid": "AMZN",
              },
              Object {
                "amount": 340,
                "avgPrice": 132,
                "lastDayRevenue": 0.52,
                "totalMarketValue": 2084,
                "totalRevenue": 0.54,
                "uid": "AAPL",
              },
            ],
            "totalMarketValue": 5000,
            "totalRevenue": -0.02,
            "uid": "Portfolio1",
          },
          "refreshing": false,
        },
        "totalMarketValue": 0.53,
        "totalRevenue": 0.25,
        "uid": "Portfolio1",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      colors={
        Array [
          "#004D40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderBottomWidth": 0,
              "borderColor": "#bcbbc1",
              "flex": 1,
              "flexDirection": "row",
              "padding": 14,
              "paddingLeft": 32,
              "paddingVertical": 16,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "marginLeft": 16,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#004D40",
                    "fontSize": 32,
                    "fontWeight": "bold",
                  }
                }
              >
                Portfolio1
              </Text>
            </View>
            <View
              style={
                Object {
                  "flexDirection": "column",
                  "marginLeft": 16,
                  "marginTop": 24,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#004D40",
                    "fontSize": 24,
                    "marginBottom": 0,
                    "textAlign": "justify",
                  }
                }
              >
                Total revenue (%)
              </Text>
              <Text
                style={
                  Object {
                    "color": "#54c242",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "marginBottom": 7.847058823529411,
                    "marginTop": -3.9235294117647057,
                    "textAlign": "justify",
                  }
                }
              >
                +25.00 %
              </Text>
            </View>
          </View>
          <View
            width={16}
          />
          <View
            style={
              Object {
                "alignItems": "flex-end",
                "flex": 0.5,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "column",
                  "marginLeft": 16,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#004D40",
                    "fontSize": 24,
                  }
                }
              >
                Revenue today
              </Text>
              <Text
                style={
                  Object {
                    "color": "#54c242",
                    "fontSize": 35.311764705882354,
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                +23.00 %
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`portfolio-list tests renders correctly with refreshing 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#004D40",
        "fontSize": 40,
        "textAlign": "center",
      }
    }
  >
    You don't have any portfolios
  </Text>
</View>
`;
